src/main/java/net/snowflake/client/jdbc/cloud/storage/SnowflakeAzureClient.java

```diff
@@ -652,7 +653,15 @@ private static void handleAzureException(
       if (((StorageException) ex).getHttpStatusCode() == 403) {
         // A 403 indicates that the SAS token has expired,
         // we need to refresh the Azure client with the new token
-        SnowflakeFileTransferAgent.renewExpiredToken(session, command, azClient);
+        if (session != null) {
+          SnowflakeFileTransferAgent.renewExpiredToken(session, command, azClient);
+        } else {
+          // If session is null we cannot renew the token so throw the ExpiredToken exception
+          throw new SnowflakeSQLException(
```

Comment:
What's the testing story for these client implementations? Presumably we'll be implicitly testing this via our streaming ingest upload tests at some point? I poked around a bit and didn't see anything obvious in the JDBC repo.

Reply:
I think what we can do here is add a system function (probably one of the `system$it()` functions that we can use to generate expired JWT tokens off the bat and then pass them down to test. There is also a `gen_jwt.py` script in the `RegressionTests` directory that you can use to generate expired tokens as well.